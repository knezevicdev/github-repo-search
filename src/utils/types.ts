export type RepositoryType = 'public' | 'private' | 'fork';
export type RepositoryTypeFilter = 'all' | RepositoryType;

export interface Repository {
  id: number;
  private: boolean;
  fork: boolean;
  name: string;
  html_url: string;
}
