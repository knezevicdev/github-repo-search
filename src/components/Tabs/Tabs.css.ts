import styled, { css } from 'styled-components';
import colors from 'styled/colors';

export const Wrapper = styled.div`
  display: flex;
  justify-content: space-around;
`;

type TabProps = {
  active?: boolean;
};

export const Tab = styled.button<TabProps>`
  background-color: transparent;
  border: 0;
  padding: 14px 5px 10px 5px;
  cursor: pointer;
  outline: none;
  position: relative;
  color: ${colors.black2};

  &::after {
    content: '';
    display: block;
    bottom: 0;
    left: 0;
    height: 1px;
    width: 100%;
    background-color: ${colors.orange1};
    position: absolute;
    opacity: 0;

    ${({ active }) =>
      active &&
      css`
        opacity: 1;
      `}
  }

  &:hover::after,
  &:active::after,
  &:focus::after {
    opacity: 1;

    ${({ active }) =>
      !active &&
      css`
        background-color: ${colors.orange2};
      `}
  }
`;
