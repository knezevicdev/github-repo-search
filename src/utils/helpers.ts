import { Repository, RepositoryType } from 'utils/types';

export const getRepositoryType = (repository: Repository): RepositoryType => {
  if (repository.fork) {
    return 'fork';
  }

  if (repository.private) {
    return 'private';
  }

  return 'public';
};
