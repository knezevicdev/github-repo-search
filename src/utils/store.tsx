import React, { ComponentType } from 'react';
import { observable, computed, comparer, action } from 'mobx';

import { RepositoryTypeFilter, Repository } from 'utils/types';
import { getRepositoryType } from 'utils/helpers';

const storeContext = React.createContext<Store | null>(null);

class Store {
  @observable isLoading = true;
  @observable hasError = false;
  @observable private allRepositories: Repository[] = [];
  @observable searchText = '';
  @observable typeFitler: RepositoryTypeFilter = 'all';
  @observable showingAll = false;

  @computed({ equals: comparer.shallow })
  private get filteredRepositories(): Repository[] {
    const searchText = this.searchText.trim().toLowerCase();
    let repositories: Repository[] = this.allRepositories;

    if (this.typeFitler !== 'all' || searchText !== '') {
      repositories = repositories.filter((repository) => {
        if (this.typeFitler !== 'all' && this.typeFitler !== getRepositoryType(repository)) return false;
        if (repository.name.toLowerCase().indexOf(searchText) === -1) return false;

        return true;
      });

      if (searchText !== '') {
        const reg = new RegExp(searchText, 'gi');

        repositories = repositories.map((repository: Repository) => ({
          ...repository,
          name: repository.name.replace(reg, (str: string) => `<em>${str}</em>`),
        }));
      }
    }

    return repositories;
  }

  @computed({ equals: comparer.shallow })
  get repositories() {
    if (!this.showingAll) return this.filteredRepositories.slice(0, 3);

    return this.filteredRepositories;
  }

  @computed
  get repositoriesCount() {
    return this.filteredRepositories.length;
  }

  @action.bound
  async fetchRepositories(githubUser: string) {
    try {
      const repositories = await fetch(`https://api.github.com/users/${githubUser}/repos`);
      this.allRepositories = await repositories.json();
    } catch {
      this.hasError = true;
    }

    this.isLoading = false;
  }

  @action.bound
  setTypeFilter(type: RepositoryTypeFilter) {
    this.typeFitler = type;
  }

  @action.bound
  setSearchText(searchText: string) {
    this.searchText = searchText;
  }

  @action.bound
  toggleShowingAll() {
    this.showingAll = !this.showingAll;
  }
}

export const StoreProvider: React.FC = ({ children }) => {
  const store = new Store();

  return <storeContext.Provider value={store}>{children}</storeContext.Provider>;
};

export const withStore = (WrappedComponent: ComponentType): React.FC => {
  const StoreHOC: React.FC = (props) => {
    return (
      <StoreProvider>
        <WrappedComponent {...props} />
      </StoreProvider>
    );
  };

  return StoreHOC;
};

export const useStore = (): Store => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
