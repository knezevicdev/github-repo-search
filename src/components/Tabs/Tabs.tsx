import React from 'react';
import { observer } from 'mobx-react';
import { action } from 'mobx';

import { Wrapper, Tab } from './Tabs.css';
import { useStore } from 'utils/store';
import { RepositoryTypeFilter } from 'utils/types';

const tabs = {
  all: 'All',
  public: 'Public',
  private: 'Private',
  fork: 'Forks',
};

const Tabs: React.FC = () => {
  const { typeFitler, setTypeFilter } = useStore();

  const onTabClick = (tab: string) => () => {
    if (tab !== typeFitler) {
      setTypeFilter(tab as RepositoryTypeFilter);
    }
  };

  return (
    <Wrapper>
      {Object.entries(tabs).map(([key, label]) => (
        <Tab key={`tab#${key}`} active={key === typeFitler} onClick={onTabClick(key)}>
          {label}
        </Tab>
      ))}
    </Wrapper>
  );
};

export default observer(Tabs);
