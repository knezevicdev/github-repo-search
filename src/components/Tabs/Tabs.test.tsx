import React from 'react';
import { shallow } from 'enzyme';

import { StoreProvider } from 'utils/store';

import Tabs from './Tabs';

describe('Tabs', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(
      <StoreProvider>
        <Tabs />
      </StoreProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
