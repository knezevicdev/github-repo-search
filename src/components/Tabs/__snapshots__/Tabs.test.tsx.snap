// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders snapshots 1`] = `
<ContextProvider
  value={
    Store {
      "allRepositories": Array [],
      "hasError": false,
      "isLoading": true,
      "searchText": "",
      "showingAll": false,
      "typeFitler": "all",
    }
  }
>
  <Tabs />
</ContextProvider>
`;
